Simplified category form

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplifies category form and refactors category saving, by adding more abstract classes. This makes sure that there is less repeated code between root and normal category management. Root category form type was already abstracted so there is little reason why form handlers shouldn't be abstracted as well.
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  | See below

## How to test

Simplifying Sell -> Catalog -> Category. Everything must look the same.

Please test the "add menu thumbnail images" capacity: add multiple, remove, and there is a limit (3) above which **the input is disabled**.

The description should have clean HTML(no javascript). Some fields can't exceed a certain number of characters meta title and meta keywords: 255, meta description: 512, name and link rewrite: 128 characters

## :notebook: BC Break
Backwards compatibility break introduced because AbstractCategoryType has changed parameters. This means that if any module extends this type(or any type that extends it) that module might crash upon updating to PS version where this update is present.

Backward comptability break introduced because RootCategoryFormDataHandler createAddCategoryCommand and createEditCategoryCommand functions are now protected. They are now protected because they were previously protected in CategoryFormDataHandler
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/21243)
<!-- Reviewable:end -->

matks commented:
Unhappy PHP CS Fixer 😄

matks commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,34 +71,61 @@ abstract class AbstractCategoryType extends TranslatorAwareType
+        if (null !== $options['id_category']) {
+            $categoryId = $options['id_category'];
+            /** @var EditableCategory $editableCategory */
+            $editableCategory = $this->queryBus->handle(new GetCategoryForEditing((int) $categoryId));
```
In theory, Queries are meant to be used in only 1 place. `GetCategoryForEditing` is a Query that fetches and return all the data needed for merchant to use an interface aimed at editing the category.

I think we should rather fetch this `canContainMoreMenuThumbnails` by another way:
- we can either use a Read Model like a DataProvider to fetch it
- or we can use the `$options` array to pass this data (see below)

In Controller, we already perform `$this->queryBus->handle(new GetCategoryForEditing((int) $categoryId));` . Then we could extract `canContainMoreMenuThumbnails` and inject it into the `$options` to be given to this form.

matks replied:
Actually it completely make sense as we already go into this `if` following an option.

So `$options['disableMenuThumbnailsUpload']` looks good to me

JevgenijVisockij replied:
Adding is to options is great idea! Thanks, will do that.

matks commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -122,19 +155,45 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                            '%displayed_label%' => $this->trans('Displayed', 'Admin.Global')
+                        ]
+                    ) . '<br>' .
+                    $this->trans(
```
I wonder if we should create a Helper class to handle these. This
```
'[1]' => '<a href="' . $this->router->generate('admin_module_manage') .  '" target="_blank" rel="noopener noreferrer nofollow">',
 '[/1]' => '</a>
 ```
 looks weird.
 
 Something like `HelperTrans::wrapTransIntoATag()` ?

JevgenijVisockij replied:
I don't see how you would do this without separating 
`"If you want a category to appear in the menu of your shop, go to [1]Modules > Module Manager[/1] and configure your menu module."` into 2 separate translatable strings. ("Modules > Module Manager" and "If you want a category to appear in the menu of your shop, go to %module_manager% and configure your menu module.")

And if you do separate it into 2 strings then I don't think helper is needed as you can just:
`$string = '<a href="' . $this->router->generate('admin_module_manage') . '" target="_blank" rel="noopener noreferrer nofollow">' . $this->trans('Modules > Module Manager', 'Dom') . '</a>';`
And pass that variable to the translation.

I agree it looks weird, but I don't think there is a better way to add HTML tags and keep it as one translatable string.

matks commented:
@JevgenijVisockij the manual `form_row` is because there is a specific display to map? It was not possible to use a single `form_widget` ?

JevgenijVisockij committed:
CS Fixer fixes

JevgenijVisockij commented:
> @JevgenijVisockij the manual `form_row` is because there is a specific display to map? It was not possible to use a single `form_widget` ?

Yes exactly. 

If I want to maintain the same the same order of inputs I need to put all form rows in correct order. 
And I need custom HTML for couple of rows to display alerts or seo preview.

JevgenijVisockij committed:
Changed the way disable_menu_thumbnails_upload is used in form.

JevgenijVisockij committed:
Fixed getting an exception when menu thumbnails is at maximum capacity

JevgenijVisockij committed:
Removing help from description as it's not correct

JevgenijVisockij commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -104,14 +104,17 @@ public function update($categoryId, array $data)
         $command = $this->createEditCategoryCommand($categoryId, $data);
 
         $this->commandBus->handle($command);
-
```
You can only add a certain amount of menuThumbnailImages. If you are at that amount warning will be displayed that you are at maximum. But you can still add another file.

So I thought it would make sense to disable form input as well since you shouldn't be adding more images anyway(and if you delete one the images page will be reloaded). 
But if I disable input it means menu_thumbnail_images won't be passed to CategoryFormDataHandler anymore, thus I add default value here. 

I tried to do it with readonly instead of disabled but it didn't disable the file input

kpodemski replied:
@JevgenijVisockij you can hack it like this:

```
input[type=file][readonly] {
    pointer-events: none;
}
```

but still available if you tab into it, if you still have this issue we can ask @NeOMakinG for advice

JevgenijVisockij committed:
Added comment to make changes clearer

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -108,10 +133,9 @@ public function buildForm(FormBuilderInterface $builder, array $options)
-            ->add('description', TranslateType::class, [
+            ->add('description', TranslatableType::class, [
+                'label' => $this->trans('Description', 'Admin.Global'),
                 'type' => FormattedTextareaType::class,
```
Note that I removed the help text "Invalid characters: <>;=#{}" for Description.
That help text doesn't make sense because the only constraint here is CleanHtml which will check for script tags and such but will allow "<>;=#{}". 

I don't know if new help text is needed to reflect that.

Julievrz replied:
You are right, if these characters are allowed, we should get rid of the help text. I don't think it is necessary to have a help text at all in that case. ☺️

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -214,6 +269,8 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('link_rewrite', TranslatableType::class, [
+                'label' => $this->trans('Friendly URL', 'Admin.Global'),
+                'help' => $this->trans('Only letters, numbers, underscore (_) and the minus (-) character are allowed.', 'Admin.Catalog.Help'),
                 'type' => TextType::class,
```
The help text here is only correct if "Accented URL" (in Traffic & SEO) is turned off. I think we should have either 

1. 2 different help texts depending on if the option turned on or not. 
2. Modify help text to let user know about the setting and what it changes.

I like the second option more as it would make it clear that Accented URL's changes the way this field is validated.

Julievrz replied:
What do you think of the following wording? (I followed your second idea)

Unless the 'Accented URL' option is enabled (in Shop parameters > Traffic & SEO), only letters, numbers, underscores (_), and hyphens (-) are allowed.

JevgenijVisockij replied:
Makes sense to me. Added your suggestion.

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -122,19 +146,45 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ])
             ->add('active', SwitchType::class, [
+                'label' => $this->trans('Displayed', 'Admin.Global'),
+                /* IMO help here is not exactly clear. I don't understand what click on "displayed" means, maybe previously it was clickable but now isin't? */
```
IMO help text (Click on "%displayed_label%" to index the category on your shop. where %displayed_label% = Dispalyed) for active is not exactly clear. I don't understand what click on "displayed" means, maybe previously it was clickable but now isin't?

Julievrz replied:
You're right. What do you think of replacing "Displayed" with "Enabled" and getting rid of the help text? I think the label is already quite clear on its own. 

![image](https://user-images.githubusercontent.com/70583503/103076621-6f6e2100-45ce-11eb-99ac-b6d11920ad4f.png)

JevgenijVisockij replied:
Makes sense to me. Implemented the changes.

Julievrz replied:
@JevgenijVisockij I can't see the string 'Displayed' anymore. Is it normal?

JevgenijVisockij commented:
Could anyone responsible for the wording check those 3 questions?
https://github.com/PrestaShop/PrestaShop/pull/22473#discussion_r547878641
https://github.com/PrestaShop/PrestaShop/pull/22473#discussion_r547865863
https://github.com/PrestaShop/PrestaShop/pull/22473#discussion_r547863221

Julievrz commented:
> Wording was already validated but I have some questions regarding wording, could someone responsible for the wording check those 3?
> [#22473 (comment)](https://github.com/PrestaShop/PrestaShop/pull/22473#discussion_r547878641)
> [#22473 (comment)](https://github.com/PrestaShop/PrestaShop/pull/22473#discussion_r547865863)
> [#22473 (comment)](https://github.com/PrestaShop/PrestaShop/pull/22473#discussion_r547863221)

Hi, thanks for your comment! I'll have a look. :)

JevgenijVisockij committed:
Changed help text for category link_rewerite to the one that makes more sense

JevgenijVisockij committed:
Removed no longer needed help text

JevgenijVisockij committed:
Fixed problems with root category

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,34 +68,58 @@ abstract class AbstractCategoryType extends TranslatorAwareType
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $disableMenuThumbnailsUpload = false;
+        if (isset($options['id_category']) && null !== $options['id_category']) {
```
```suggestion
        if (isset($options['id_category'])) {
```

Progi1984 replied:
isset : `Determine if a variable is considered set, this means if a variable is declared and is different than null.`

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

Co-authored-by: Progi1984 <progi1984@gmail.com>

Julievrz commented:
Thanks for your advice on the wording, I changed the label.😊

matks commented:
@JevgenijVisockij can you please improve the "How to test" of the PR 😉 I started the improvement but it would be great to indicate, for QA team, the limit above which input is disabled so they can test it.

khouloudbelguith commented:
Hi @JevgenijVisockij, @matks,

This PR is ready for QA?

Thank you!

khouloudbelguith commented:
Hi @JevgenijVisockij,

After a check, the` Category cover image` cannot be uploaded and the `Category thumbnail` can not be uploaded.
I attached a screen record
https://drive.google.com/file/d/1Q-9HZMyY2pY1cTWaFMB3qeJiPbwdF7ZD/view?usp=sharing

Thank you!

JevgenijVisockij commented:
> @JevgenijVisockij can you please improve the "How to test" of the PR wink I started the improvement but it would be great to indicate, for QA team, the limit above which input is disabled so they can test it.

Good point. Updated the description ;)

JevgenijVisockij commented:
> Hi @JevgenijVisockij,
> 
> After a check, the` Category cover image` cannot be uploaded and the `Category thumbnail` can not be uploaded.
> I attached a screen record
> https://drive.google.com/file/d/1Q-9HZMyY2pY1cTWaFMB3qeJiPbwdF7ZD/view?usp=sharing
> 
> Thank you!

Thanks for Q&A fixed the issue and updated some other things in PR.

This is ready for reviews.

Julievrz commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,51 +70,87 @@ abstract class AbstractCategoryType extends TranslatorAwareType
+                        new Length([
+                            'max' => self::MAX_TITLE_LENGTH,
+                            'maxMessage' => $this->trans(
+                                'This field cannot be longer than %limit% characters',
```
```suggestion
                                'This field cannot be longer than %limit% characters.',
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -198,36 +257,67 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                        new Length([
+                            'max' => SeoSettings::MAX_KEYWORDS_LENGTH,
+                            'maxMessage' => $this->trans(
+                                'This field cannot be longer than %limit% characters',
```
```suggestion
                                'This field cannot be longer than %limit% characters.',
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -198,36 +257,67 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                        new Length([
+                            'max' => self::MAX_TITLE_LENGTH,
+                            'maxMessage' => $this->trans(
+                                'This field cannot be longer than %limit% characters',
```
```suggestion
                                'This field cannot be longer than %limit% characters.',
```

Julievrz reviewed (commented):
A small wording suggestion that I had missed before approving it for good. Thank you!

JevgenijVisockij committed:
Fixed category images not showing up

Progi1984 commented on src/Core/Domain/Category/SeoSettings.php:
```diff
@@ -50,4 +50,9 @@ final class SeoSettings
+    /**
+     * Maximum length of SEO keywords (value is constrained by database)
+     */
+    const MAX_KEYWORDS_LENGTH = 255;
```
```suggestion
    public const MAX_KEYWORDS_LENGTH = 255;
```

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

Added dot constraint message

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

Added dot to constraint message.

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php

Added dot to constraint message.

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Domain/Category/SeoSettings.php

Added public to constant definition

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 commented:
@JevgenijVisockij Ready for QA ?

khouloudbelguith commented:
Hi @JevgenijVisockij,

I have some issues:

1. Install shop with English language & France country
2. In the BO > Catalog > Catgeories > Add new category page
3. The `Id parent` expression is NOK
**With PR** 
![image](https://user-images.githubusercontent.com/16067358/106129650-bd8fba80-6160-11eb-881a-23f6eb98e2b3.png)
**Without PR**
![image](https://user-images.githubusercontent.com/16067358/106129801-e44df100-6160-11eb-88cf-43141a33cbce.png)

I found some issues reproduced with or without PR (1771 => ok)
1.  In the BO > Catalog > Catgeories > Add new category page
2. In the description section > Add this text `<>`
3. Save > see the error 
![image](https://user-images.githubusercontent.com/16067358/106130243-68a07400-6161-11eb-9081-4ae6db0d5fcf.png)


1. In the BO > Catalog > Categories > Add new category
2. In the Description field > Click on `Bullet list` & `Numbered list`
3. See the error in the console 
![image](https://user-images.githubusercontent.com/16067358/106132772-a94dbc80-6164-11eb-9f5a-df6d784ef693.png)

Reproduced with or without PR (1771 => ok)
1. Disable debug mode
1. In the BO > Catalog > Categories
2. In the image files => add incompatible format image (.iso, .webp, .ico, .pdf, .zip)
3. Save
4. See error => unclear alert 
![image](https://user-images.githubusercontent.com/16067358/106134692-30039900-6167-11eb-874f-7e72d28fb7bf.png)

Reproduced with or without PR (1771 => ok)

1.Disbale debug mode
1. Enable multistore 
2. Create a new shop
3. Go to BO > Catalog > Categories page (shop 1 selected)
4. Create a root category > save
5. You are in the Category root list page 
![image](https://user-images.githubusercontent.com/16067358/106135186-dd76ac80-6167-11eb-834c-062ce357927f.png)
6. Add new category
7. Set only the name & save
8. See error 
![image](https://user-images.githubusercontent.com/16067358/106135292-01d28900-6168-11eb-8584-22f5ec68f386.png)


About this:
>Please test the "add menu thumbnail images" capacity: add multiple, remove, and there is a limit (3) above which the input is

0. disable debug mode
1. I can select more than 3 images in the same time 
![image](https://user-images.githubusercontent.com/16067358/106135606-71e10f00-6168-11eb-9df3-b5334a6a6be0.png)
2. Save
3. See error 
![image](https://user-images.githubusercontent.com/16067358/106135942-dc924a80-6168-11eb-955f-3ebf349bc53a.png)


Thank you!

JevgenijVisockij committed:
Added label for parent category

JevgenijVisockij committed:
Added maximum values to certain category fields

# Conflicts:
#	src/Core/Domain/Category/SeoSettings.php

JevgenijVisockij committed:
Added proper error throwing for multiple images in menu thumbnails

JevgenijVisockij committed:
CS fixer fixes

Progi1984 commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -81,16 +83,26 @@ public function __construct(
+        /**
+         * In some cases in form menu_thumbnail_images can be disabled so value won't get here.
+         */
+        $menuThumbnailImages = [];
+        if (isset($data['menu_thumbnail_images'])) {
+            $menuThumbnailImages = $data['menu_thumbnail_images'];
+        }
```
```suggestion
        $menuThumbnailImages = $data['menu_thumbnail_images'] ?? [];
```

Progi1984 commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -101,17 +113,29 @@ public function create(array $data)
+        /**
+         * In some cases in form menu_thumbnail_images can be disabled so value won't get here.
+         */
+        $menuThumbnailImages = [];
+        if (isset($data['menu_thumbnail_images'])) {
+            $menuThumbnailImages = $data['menu_thumbnail_images'];
+        }
```
```suggestion
        $menuThumbnailImages = $data['menu_thumbnail_images'] ?? [];
```

Progi1984 reviewed (commented):
Some feedbacks

JevgenijVisockij committed:
Adding raw option to data column

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Shortended the syntax

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij committed:
Changed from escape to raw

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Using short if syntax

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 commented on src/Core/Grid/Column/Type/Common/HtmlColumn.php:
```diff
@@ -0,0 +1,63 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Column\Type\Common;
```
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Grid\Column\Type\Common;
```

Progi1984 commented on src/Core/Grid/Column/Type/Common/HtmlColumn.php:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Class Column defines most simple column in the grid that renders raw data.
+ */
+final class HtmlColumn extends AbstractColumn
```
Could you add a doc : https://devdocs.prestashop.com/1.7/development/components/grid/columns-reference/  ?

JevgenijVisockij replied:
All right will do that.

Progi1984 replied:
Wait for your PR

JevgenijVisockij replied:
https://github.com/PrestaShop/docs/pull/879

Progi1984 reviewed (commented):
Some feedbacks

JevgenijVisockij committed:
Added bugfix in case you are adding category from root

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij commented:
> Hi @JevgenijVisockij,
> 
> I have some issues:
> 
>     1. Install shop with English language & France country
> 
>     2. In the BO > Catalog > Catgeories > Add new category page
> 
>     3. The `Id parent` expression is NOK
>        **With PR**
>        ![image](https://user-images.githubusercontent.com/16067358/106129650-bd8fba80-6160-11eb-881a-23f6eb98e2b3.png)
>        **Without PR**
>        ![image](https://user-images.githubusercontent.com/16067358/106129801-e44df100-6160-11eb-88cf-43141a33cbce.png)
> 
> 
> I found some issues reproduced with or without PR (1771 => ok)
> 
>     1. In the BO > Catalog > Catgeories > Add new category page
> 
>     2. In the description section > Add this text `<>`
> 
>     3. Save > see the error
>        ![image](https://user-images.githubusercontent.com/16067358/106130243-68a07400-6161-11eb-9081-4ae6db0d5fcf.png)
> 
>     4. In the BO > Catalog > Categories > Add new category
> 
>     5. In the Description field > Click on `Bullet list` & `Numbered list`
> 
>     6. See the error in the console
>        ![image](https://user-images.githubusercontent.com/16067358/106132772-a94dbc80-6164-11eb-9f5a-df6d784ef693.png)
> 
> 
> Reproduced with or without PR (1771 => ok)
> 
>     1. Disable debug mode
> 
>     2. In the BO > Catalog > Categories
> 
>     3. In the image files => add incompatible format image (.iso, .webp, .ico, .pdf, .zip)
> 
>     4. Save
> 
>     5. See error => unclear alert
>        ![image](https://user-images.githubusercontent.com/16067358/106134692-30039900-6167-11eb-874f-7e72d28fb7bf.png)
> 
> 
> Reproduced with or without PR (1771 => ok)
> 
> 1.Disbale debug mode
> 
>     1. Enable multistore
> 
>     2. Create a new shop
> 
>     3. Go to BO > Catalog > Categories page (shop 1 selected)
> 
>     4. Create a root category > save
> 
>     5. You are in the Category root list page
>        ![image](https://user-images.githubusercontent.com/16067358/106135186-dd76ac80-6167-11eb-834c-062ce357927f.png)
> 
>     6. Add new category
> 
>     7. Set only the name & save
> 
>     8. See error
>        ![image](https://user-images.githubusercontent.com/16067358/106135292-01d28900-6168-11eb-8584-22f5ec68f386.png)
> 
> 
> About this:
> 
> > Please test the "add menu thumbnail images" capacity: add multiple, remove, and there is a limit (3) above which the input is
> 
>     1. disable debug mode
> 
>     2. I can select more than 3 images in the same time
>        ![image](https://user-images.githubusercontent.com/16067358/106135606-71e10f00-6168-11eb-9df3-b5334a6a6be0.png)
> 
>     3. Save
> 
>     4. See error
>        ![image](https://user-images.githubusercontent.com/16067358/106135942-dc924a80-6168-11eb-955f-3ebf349bc53a.png)
> 
> 
> Thank you!

Hey, fixed all problems aside for click on Bullet list & Numbered list console error as it's the same problem as in CMS page, so there is a separate issue for it.

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -162,6 +162,14 @@ public function createAction(Request $request)
+         * Parent category can be root category only if you specifically click to add root category.
+         * In all other cases it should be at least home category(Or one of it's children).
+         */
+        if ($parentId === $this->configuration->getInt('PS_ROOT_CATEGORY')) {
+            $parentId = $this->configuration->getInt('PS_HOME_CATEGORY');
```
```suggestion
        $configRootCategory = $this->configuration->getInt('PS_ROOT_CATEGORY');
        if ($parentId === $configRootCategory) {
            $parentId = $configRootCategory;
```

Progi1984 reviewed (commented):
Two feedbacks

Julievrz reviewed (dismissed):
The wording is good, thank you.

matks committed:
Apply suggestions from code review

Co-authored-by: Progi1984 <progi1984@gmail.com>

PierreRambaud commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -81,16 +83,23 @@ public function __construct(
      */
     public function create(array $data)
     {
+        if (count($data['menu_thumbnail_images']) > count(MenuThumbnailId::ALLOWED_ID_VALUES)) {
```
```suggestion
        if (!isset($data['menu_thumbnail_images']) || count($data['menu_thumbnail_images']) > count(MenuThumbnailId::ALLOWED_ID_VALUES)) {
```

PierreRambaud commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -101,17 +110,26 @@ public function create(array $data)
     {
+        $availableKeys = $this->getAvailableKeys((int) $categoryId);
+
+        if (count($data['menu_thumbnail_images']) > count($availableKeys)) {
```
```suggestion
        if (!isset($data['menu_thumbnail_images']) || count($data['menu_thumbnail_images']) > count($availableKeys)) {
```

PierreRambaud commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -101,17 +110,26 @@ public function create(array $data)
+        $availableKeys = $this->getAvailableKeys((int) $categoryId);
+
+        if (count($data['menu_thumbnail_images']) > count($availableKeys)) {
+            throw new MenuThumbnailsLimitException(sprintf('Maximum number of menu thumbnails was reached for category "%s"', $categoryId));
```
```suggestion
            throw new MenuThumbnailsLimitException(sprintf('Maximum number of menu thumbnails was reached for category "%d"', $categoryId));
```

PierreRambaud commented on src/Core/Grid/Column/Type/DataColumn.php:
```diff
@@ -55,9 +55,11 @@ protected function configureOptions(OptionsResolver $resolver)
             ])
             ->setDefaults([
                 'clickable' => true,
+                'raw' => false,
```
I think you can revert this change because of `HtmlColumn`

atomiix replied:
Didn't see that! 😅

PierreRambaud commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -826,6 +838,16 @@ private function getErrorMessages()
             ),
+            UploadedImageConstraintException::class => [
+                UploadedImageConstraintException::EXCEEDED_SIZE => $this->trans(
+                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error',
```
```suggestion
                    'Max file size allowed is "%d" bytes.', 'Admin.Notifications.Error',
```

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -52,6 +53,8 @@
  */
 abstract class AbstractCategoryType extends TranslatorAwareType
 {
+    private const MAX_TITLE_LENGTH = 128;
```
Any reason to use private? Cannot be used in children classes :(

PierreRambaud replied:
Seems to be a duplicate from SeoSettings::MAX_TITLE_LENGTH

JevgenijVisockij replied:
Yeah it should be public fixed that, but it's not duplicate. The one from SeoSettings is meant for meta title and has different length then actual title.

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,51 +70,87 @@ abstract class AbstractCategoryType extends TranslatorAwareType
+                                'This field cannot be longer than %limit% characters.',
+                                'Admin.Notifications.Error',
+                                [
+                                    '%limit%' => SeoSettings::MAX_TITLE_LENGTH,
```
```suggestion
                                    '%limit%' => static::MAX_TITLE_LENGTH,
```
Otherwise there is something wrong with the `MAX_TITLE_LENGTH` usage

PierreRambaud commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -198,36 +257,67 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                                'This field cannot be longer than %limit% characters.',
+                                'Admin.Notifications.Error',
+                                [
+                                    '%limit%' => SeoSettings::MAX_TITLE_LENGTH,
```
same as previous

PierreRambaud reviewed (changes requested):
Good job, few comments remaining and it's ok for me :)

PierreRambaud commented:
I'm sorry, I remove the Waiting for QA label, at least because of DataColumn, it doesn't require to ass a raw parameter :)
There are also problems with consts :sweat_smile:

JevgenijVisockij committed:
Replaced self with static

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Checking isset for menu_thumbnails_images

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Checking for isset for menu thumbnail images

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

Using as decimal in sprintf instead of string

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Removed raw from dataColumn as it's no longer needed

JevgenijVisockij committed:
Update src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php

Using decimal instead of string for sprintf

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Making MAX_TITLE_LENGTH public

JevgenijVisockij committed:
Making constraint for max title length instead of SeoSettings max title length

JevgenijVisockij commented:
> I'm sorry, I remove the Waiting for QA label, at least because of DataColumn, it doesn't require to ass a raw parameter :)
> There are also problems with consts sweat_smile

Well Code Quality above all else ^^. I have addressed your feedback.

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -162,6 +162,15 @@ public function createAction(Request $request)
+         * In all other cases it should be at least home category(Or one of it's children).
+         */
+        $configRootCategory = $this->configuration->getInt('PS_ROOT_CATEGORY');
+        if ($parentId === $configRootCategory) {
+            $parentId = $configRootCategory;
+        }
```
If it is equals then we assign the variable. Isn't it' weird ?

JevgenijVisockij replied:
You are right. I see it was a suggestion that got weird, so I guess something got weird there and I later missed it. 

Good catch! Fixed it.

Progi1984 reviewed (changes requested):
Could you add unit tests for `src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php` ? Please

JevgenijVisockij committed:
Changed from root category to home category

matks commented:
> Could you add unit tests for `src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php` ? Please

I'd like to provide my opinion 😉  : although indeed it would be an _awesome benefit_ 🤩  for this PR to have a unit testing the logic you added about thumbnail image limitation, I see that the class is involved into file uploading features. **This can be very hard to unit test**.

If we find out that the size of the test needed to verify this behavior is way larger than the benefit it brings, **it can be scoped for another PR** as the aim of this PR was to simplify the Category form, not to introduce a test setup for a class that is able to handle dummy file uploads 😄  . Also we want to keep the size of this PR manageable.

Let's be pragmatic 💪

JevgenijVisockij commented:
> > Could you add unit tests for `src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php` ? Please
> 
> I'd like to provide my opinion wink : although indeed it would be an _awesome benefit_ star_struck for this PR to have a unit testing the logic you added about thumbnail image limitation, I see that the class is involved into file uploading features. **This can be very hard to unit test**.
> 
> If we find out that the size of the test needed to verify this behavior is way larger than the benefit it brings, **it can be scoped for another PR** as the aim of this PR was to simplify the Category form, not to introduce a test setup for a class that is able to handle dummy file uploads smile . Also we want to keep the size of this PR manageable.
> 
> Let's be pragmatic muscle

I was thinking about what could be tested here and I think it would make sense to test getAvailableKeys function.
https://github.com/PrestaShop/PrestaShop/pull/22473/files#diff-e9a847db7e2cb76ff119cca4a5b976eb83d937fa7d0f89dcedc417af0ae8b0e2

Problem is that it has scandir and after little research I couldn't find to override how it works for testing purposes.
```$files = scandir(_PS_CAT_IMG_DIR_, SCANDIR_SORT_NONE);```
But if this line was separated from the whole function and ```$files``` was passed as parameter, the whole function would become easy to test. 
WDYT?

JevgenijVisockij committed:
Adding test for getAvailableKeys function

Progi1984 commented on tests/Unit/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandlerTest.php:
```diff
@@ -0,0 +1,73 @@
+            $imageUploaderMock
+        );
+
+        $reflectionClass = new \ReflectionClass(CategoryFormDataHandler::class);
```
Move `\ReflectionClass` to use section.

Progi1984 commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -198,4 +216,25 @@ private function uploadImages(
+    /**
+     * @param int $categoryId
+     *
+     * @return array
```
```suggestion
     * @return array<int, int>
```

Progi1984 commented on tests/Unit/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandlerTest.php:
```diff
@@ -0,0 +1,73 @@
+use PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataHandler\CategoryFormDataHandler;
+use PrestaShop\PrestaShop\Core\Image\Uploader\ImageUploaderInterface;
+
+class CategoryFormDataHandlerTest extends TestCase
```
:heart_eyes:

jolelievre replied:
I'm not sure to understand the purpose of this test, testing private methods has very little interests 🤔

Progi1984 reviewed (commented):
Small feedback

JevgenijVisockij committed:
Fixed some issues with root category by aligning it closer with category

JevgenijVisockij committed:
Moved most of the category handler code to abstract handler

JevgenijVisockij committed:
CS fixer fix

JevgenijVisockij committed:
Implemented feedback for getAvailableKeys function and its test

JevgenijVisockij commented:
> Small feedback

Thanks! Impelmented it.

Also I made some changes to category form handler, because root category editing didn't work well for me. 

Now there is abstract category form handler, because handling for root category and normal category is nearly identical aside for creation of command.

JevgenijVisockij committed:
Removed constant changing

JevgenijVisockij committed:
Added abstract functions

JevgenijVisockij committed:
Fixed abstract category handler

Progi1984 commented on src/Core/Form/IdentifiableObject/DataHandler/AbstractCategoryFormDataHandler.php:
```diff
@@ -0,0 +1,183 @@
+        return array_diff(MenuThumbnailId::ALLOWED_ID_VALUES, $usedKeys);
+    }
+
+    abstract protected function createAddCategoryCommand(array $data);
+
+    abstract protected function createEditCategoryCommand($categoryId, array $data);
```
Return values is missing
Type for $categoryId is missing

JevgenijVisockij replied:
Abstracted add/addroot and edit/editroot category commands so return value can be added.

Progi1984 commented on src/Core/Form/IdentifiableObject/DataHandler/AbstractCategoryFormDataHandler.php:
```diff
@@ -0,0 +1,183 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function update($categoryId, array $data)
```
Add type for $categoryId

JevgenijVisockij replied:
This implements FormDataHandlerInterface interface, which doesn't have type so can't change it.

Progi1984 reviewed (commented):
Small comments on type hint and it's good

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -831,6 +846,16 @@ private function getErrorMessages()
             ),
+            UploadedImageConstraintException::class => [
+                UploadedImageConstraintException::EXCEEDED_SIZE => $this->trans(
+                    'Max file size allowed is "%d" bytes.', 'Admin.Notifications.Error',
```
```suggestion
                    'Max file size allowed is "%s" bytes.', 'Admin.Notifications.Error',
```

Progi1984 replied:
@Julievrz No, the file size is an integer not a string.

Julievrz replied:
@Progi1984 Thanks.  I guess the existing string was wrong too then. 
![image](https://user-images.githubusercontent.com/70583503/113157720-734bb800-923b-11eb-9564-11035e42f689.png)

Progi1984 replied:
Approved ;)

JevgenijVisockij committed:
Added abstract root command to allow for typehinting

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Added int typehint for categoryId

JevgenijVisockij committed:
Changed categoryId so it get's converted to int earleir

JevgenijVisockij commented:
> Small comments on type hint and it's good

Addressed them.

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedNames()
```
```suggestion
    public function getLocalizedNames(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function setLocalizedNames(array $localizedNames)
```
```suggestion
    public function setLocalizedNames(array $localizedNames): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedLinkRewrites()
```
```suggestion
    public function getLocalizedLinkRewrites(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function setLocalizedLinkRewrites(array $localizedLinkRewrites)
```
```suggestion
    public function setLocalizedLinkRewrites(array $localizedLinkRewrites): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedDescriptions()
```
```suggestion
    public function getLocalizedDescriptions(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @return $this
+     */
+    public function setLocalizedDescriptions(array $localizedDescriptions)
```
```suggestion
    public function setLocalizedDescriptions(array $localizedDescriptions): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return bool
+     */
+    public function isActive()
```
```suggestion
    public function isActive(): bool
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function setIsActive($isActive)
```
```suggestion
    public function setIsActive(bool $isActive): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedMetaTitles()
```
```suggestion
    public function getLocalizedMetaTitles(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @return $this
+     */
+    public function setLocalizedMetaTitles(array $localizedMetaTitles)
```
```suggestion
    public function setLocalizedMetaTitles(array $localizedMetaTitles): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedMetaDescriptions()
```
```suggestion
    public function getLocalizedMetaDescriptions(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @return $this
+     */
+    public function setLocalizedMetaDescriptions(array $localizedMetaDescriptions)
```
```suggestion
    public function setLocalizedMetaDescriptions(array $localizedMetaDescriptions): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedMetaKeywords()
```
```suggestion
    public function getLocalizedMetaKeywords(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @return $this
+     */
+    public function setLocalizedMetaKeywords(array $localizedMetaKeywords)
```
```suggestion
    public function setLocalizedMetaKeywords(array $localizedMetaKeywords): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @return $this
+     */
+    public function setAssociatedGroupIds(array $associatedGroupIds)
```
```suggestion
    public function setAssociatedGroupIds(array $associatedGroupIds): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return int[]
+     */
+    public function getAssociatedGroupIds()
```
```suggestion
    public function getAssociatedGroupIds(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+    /**
+     * @return int[]
+     */
+    public function getAssociatedShopIds()
```
```suggestion
    public function getAssociatedShopIds(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+     *
+     * @return $this
+     */
+    public function setAssociatedShopIds(array $associatedShopIds)
```
```suggestion
    public function setAssociatedShopIds(array $associatedShopIds): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Category\Command;
```
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\Category\Command;
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,293 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Category\Command;
```
```suggestion
declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\Category\Command;
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @param int $categoryId
+     */
+    public function __construct($categoryId)
```
```suggestion
    public function __construct(int $categoryId)
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return CategoryId
+     */
+    public function getCategoryId()
```
```suggestion
    public function getCategoryId(): CategoryId
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedNames()
```
```suggestion
    public function getLocalizedNames(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function setLocalizedNames(array $localizedNames)
```
```suggestion
    public function setLocalizedNames(array $localizedNames): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedLinkRewrites()
```
```suggestion
    public function getLocalizedLinkRewrites(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function setLocalizedLinkRewrites(array $localizedLinkRewrites)
```
```suggestion
    public function setLocalizedLinkRewrites(array $localizedLinkRewrites): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedDescriptions()
```
```suggestion
    public function getLocalizedDescriptions(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @return $this
+     */
+    public function setLocalizedDescriptions(array $localizedDescriptions)
```
```suggestion
    public function setLocalizedDescriptions(array $localizedDescriptions): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return bool
+     */
+    public function isActive()
```
```suggestion
    public function isActive(): bool
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @throws CategoryConstraintException
+     */
+    public function setIsActive($isActive)
```
```suggestion
    public function setIsActive(bool $isActive): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedMetaTitles()
```
```suggestion
    public function getLocalizedMetaTitles(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @return $this
+     */
+    public function setLocalizedMetaTitles(array $localizedMetaTitles)
```
```suggestion
    public function setLocalizedMetaTitles(array $localizedMetaTitles): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedMetaDescriptions()
```
```suggestion
    public function getLocalizedMetaDescriptions(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @return $this
+     */
+    public function setLocalizedMetaDescriptions(array $localizedMetaDescriptions)
```
```suggestion
    public function setLocalizedMetaDescriptions(array $localizedMetaDescriptions): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return string[]
+     */
+    public function getLocalizedMetaKeywords()
```
```suggestion
    public function getLocalizedMetaKeywords(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @return $this
+     */
+    public function setLocalizedMetaKeywords(array $localizedMetaKeywords)
```
```suggestion
    public function setLocalizedMetaKeywords(array $localizedMetaKeywords): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return int[]
+     */
+    public function getAssociatedGroupIds()
```
```suggestion
    public function getAssociatedGroupIds(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @return $this
+     */
+    public function setAssociatedGroupIds(array $associatedGroupIds)
```
```suggestion
    public function setAssociatedGroupIds(array $associatedGroupIds): self
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+    /**
+     * @return int[]
+     */
+    public function getAssociatedShopIds()
```
```suggestion
    public function getAssociatedShopIds(): array
```

Progi1984 commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,300 @@
+     *
+     * @return $this
+     */
+    public function setAssociatedShopIds(array $associatedShopIds)
```
```suggestion
    public function setAssociatedShopIds(array $associatedShopIds): self
```

Progi1984 reviewed (commented):
Some type hint

JevgenijVisockij commented:
> Some type hint

Those typehints would have broken backwards comptability that's why I didn't add them in the first place. But I guess so would some of my changes. Will add them thanks!

JevgenijVisockij committed:
Apply suggestions from code review

Adding typehints

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Update src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php

Adding strict types

Co-authored-by: Progi1984 <progi1984@gmail.com>

Progi1984 commented:
@JevgenijVisockij Integrations Tests are :red_circle:  :(

JevgenijVisockij committed:
Changed return types so that most params can be null

JevgenijVisockij committed:
made most params in add category command nullable

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
CS fixer fixes

matks commented:
Hi @JevgenijVisockij I think this PR is no more just "Simplifies category form" 😄  can you please describe a little more what are the changes submitted in this PR? I think there is quite a lot of refactoring using abstract classes.

khouloudbelguith commented:
Hi @JevgenijVisockij,

I have an issue following these steps:
1. Go to BO > Catalog > Categories > Create new category
2. In the Menu thumbnails, add 3 valid images
3. Save
4. Edit the last created category
5. Try to edit any field (the name or description or add a cover image for example)
6. Click on save
7. See red alert => cannot update the category 
![image](https://user-images.githubusercontent.com/16067358/117036418-b5b65800-ad05-11eb-862f-a7b453cee580.png)

Thank you!

JevgenijVisockij committed:
Fixed issue with form not saving due to images when it should

JevgenijVisockij commented:
> Hi @JevgenijVisockij,
> 
> I have an issue following these steps:
> 
>     1. Go to BO > Catalog > Categories > Create new category
> 
>     2. In the Menu thumbnails, add 3 valid images
> 
>     3. Save
> 
>     4. Edit the last created category
> 
>     5. Try to edit any field (the name or description or add a cover image for example)
> 
>     6. Click on save
> 
>     7. See red alert => cannot update the category
>        ![image](https://user-images.githubusercontent.com/16067358/117036418-b5b65800-ad05-11eb-862f-a7b453cee580.png)
> 
> 
> Thank you!

Thanks for Q&A! Fixed the issue.

JevgenijVisockij committed:
Removed HTML column since it's added in another PR

JevgenijVisockij committed:
CS fixer fix

Julievrz reviewed (dismissed):
Wording ✔️

khouloudbelguith commented:
Hi @JevgenijVisockij,

This PR needs to be reviewed by our devs.

Thank you!

Progi1984 commented on admin-dev/themes/new-theme/js/pages/category/index.js:
```diff
@@ -70,11 +68,10 @@ $(() => {
   showcaseCard.addExtension(new ShowcaseCardCloseExtension());
 
-  window.prestashop.component.initComponents(['TranslatableField', 'TinyMCEEditor', 'TranslatableInput']);
+  window.prestashop.component.initComponents(['TranslatableField', 'TinyMCEEditor', 'TranslatableInput', 'TextWithRecommendedLengthCounter']);
```
ESLint warns you :) :red_circle:

Progi1984 reviewed (commented):
A small alert from ESLint

JevgenijVisockij committed:
Fixed ESLint problem

PierreRambaud commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,295 @@
+     */
+    public function setIsActive(bool $isActive): self
+    {
+        if (!is_bool($isActive)) {
+            throw new CategoryConstraintException('Invalid Category status supplied', CategoryConstraintException::INVALID_STATUS);
+        }
+
+        $this->isActive = $isActive;
```
```suggestion
        $this->isActive = $isActive;
```
It's useless and can't happen

PierreRambaud commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,302 @@
+     */
+    public function setIsActive(bool $isActive): self
+    {
+        if (!is_bool($isActive)) {
+            throw new CategoryConstraintException('Invalid Category status supplied', CategoryConstraintException::INVALID_STATUS);
+        }
+
```
```suggestion
```

PierreRambaud commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,302 @@
+     */
+    public function setAssociatedShopIds(array $associatedShopIds): self
+    {
+        $this->associatedShopIds = $associatedShopIds;
```
```suggestion
        $this->associatedShopIds = array_map('intval', $associatedShopIds);
```

Or something similar because nothing say it will contaéins only ids?

JevgenijVisockij replied:
Yeah array map makes sense thanks!

JevgenijVisockij committed:
Update src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php

Adding array map to make sure all ids are int

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php

Removed uneeded checking for bool

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Removed no lnoger needed check

JevgenijVisockij committed:
Added TextWithRecommendedLengthCounter component

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.js
#	admin-dev/themes/new-theme/js/pages/category/index.js

# Conflicts:
#	admin-dev/themes/new-theme/js/app/utils/init-components.js
#	admin-dev/themes/new-theme/js/pages/category/index.ts

atomiix commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -23,7 +23,7 @@
  *#}
 
-{% import 'PrestaShopBundle:Admin:macros.html.twig' as ps %}
+{% form_theme categoryForm 'PrestaShopBundle:Admin/TwigTemplateForm:prestashop_ui_kit.html.twig' %}
```
```suggestion
{% form_theme categoryForm '@PrestaShop/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig' %}
```

atomiix reviewed (changes requested):
Just a comment about twig files but nice work!

JevgenijVisockij committed:
Simplified category form

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig

# Conflicts:
#	templates/bundles/PrestaShopBundle/Admin/Sell/Catalog/Categories/Blocks/form.html.twig

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig

Fixed twig forms url

Co-authored-by: atomiix <tbaccelli@gmail.com>
# Conflicts:
#	templates/bundles/PrestaShopBundle/Admin/Sell/Catalog/Categories/Blocks/form.html.twig

atomiix commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -26,103 +26,24 @@
  * Creates/updates category from data submitted in category form
  */
-final class CategoryFormDataHandler implements FormDataHandlerInterface
+class CategoryFormDataHandler extends AbstractCategoryFormDataHandler
```
Why not keep it final?

JevgenijVisockij replied:
Must have been an accident added it back.

JevgenijVisockij committed:
Added new "HTML" column

# Conflicts:
#	src/Core/Grid/Column/Type/Common/HtmlColumn.php

JevgenijVisockij committed:
Fixed help link not working

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php

JevgenijVisockij committed:
Added back final

PierreRambaud commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -26,103 +26,24 @@
  * Creates/updates category from data submitted in category form
  */
-final class CategoryFormDataHandler implements FormDataHandlerInterface
+final class CategoryFormDataHandler extends AbstractCategoryFormDataHandler
```
If you use Final, put `@internal`

PierreRambaud commented on src/Core/Domain/Category/Command/AbstractAddCategoryCommand.php:
```diff
@@ -0,0 +1,291 @@
+/**
+ * Class AddRootCategoryCommand adds new root category.
+ */
+class AbstractAddCategoryCommand
```
Why call it abstract and not use abstract?
```suggestion
abstract class AbstractAddCategoryCommand
```

PierreRambaud commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,298 @@
+/**
+ * Class EditRootCategoryCommand edits given root category.
+ */
+class AbstractEditCategoryCommand
```
```suggestion
abstract class AbstractEditCategoryCommand
```
Same as previous

PierreRambaud commented on src/Core/Form/IdentifiableObject/DataHandler/RootCategoryFormDataHandler.php:
```diff
@@ -26,101 +26,24 @@
  * Creates/updates root category from data submited in category form
  */
-final class RootCategoryFormDataHandler implements FormDataHandlerInterface
+final class RootCategoryFormDataHandler extends AbstractCategoryFormDataHandler
```
same as previous about internal statement

PierreRambaud commented on src/Core/Form/IdentifiableObject/DataHandler/RootCategoryFormDataHandler.php:
```diff
@@ -26,101 +26,24 @@
      *
      * @return AddRootCategoryCommand
      */
-    public function createAddRootCategoryCommand(array $data)
```
This is a BC breaks, you have to document it :thinking:

JevgenijVisockij replied:
Updated PR desecription. Or do I need to add comment to code as well?

jolelievre replied:
It's indeed a BC break so it should documented But this method should have never been public in the first place

PierreRambaud commented on src/Core/Grid/Column/Type/Common/HtmlColumn.php:
```diff
@@ -1,65 +0,0 @@
-/**
- * Class Column defines most simple column in the grid that renders raw data.
- */
-final class HtmlColumn extends AbstractColumn
```
Why removing this column type?
:thinking:

JevgenijVisockij replied:
I think I accidentally removed it because it was added in another PR as well. Adding it back.

PierreRambaud reviewed (commented):
Few comments

sowbiba committed:
Use internal annotation for final classes + add abstract keyword

JevgenijVisockij committed:
Adding back HTML column

jolelievre commented on src/Core/Domain/Category/Command/AbstractEditCategoryCommand.php:
```diff
@@ -0,0 +1,296 @@
+/**
+ * Class EditRootCategoryCommand edits given root category.
+ */
+abstract class AbstractEditCategoryCommand
```
It feels like most of this code is common with `AbstractAddCategoryCommand` Maybe we could create a common `AbstractCategoryCommand` ?

jolelievre commented on src/Core/Domain/Category/Command/AddCategoryCommand.php:
```diff
@@ -31,73 +31,17 @@
-            ->setIsActive($isActive)
-            ->setParentCategoryId($parentCategoryId);
+        parent::__construct($localizedNames, $localizedLinkRewrites, $isActive);
+        $this->parentCategoryId = $parentCategoryId;
```
```suggestion
        $this->setParentCategoryId($parentCategoryId);
```

jolelievre replied:
Since the setter has internal checking you should use it in the constructor

jolelievre commented on src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php:
```diff
@@ -26,103 +26,29 @@
+ * @internal
  */
-final class CategoryFormDataHandler implements FormDataHandlerInterface
+final class CategoryFormDataHandler extends AbstractCategoryFormDataHandler
```
You can remove the `final` (and internal) we try to open extendability of all our services and handlers now

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -165,6 +166,15 @@ public function createAction(Request $request)
+         * Parent category can be root category only if you specifically click to add root category.
+         * In all other cases it should be at least home category(Or one of it's children).
+         */
+        $configRootCategory = $this->configuration->getInt('PS_ROOT_CATEGORY');
```
That's weird it means root category should never be selected as parent right? So how comes it can be selected as parent of this route?
Since root category have their own route and controller we shouldn't bother handling them here? Or do we?

It seems like it's fixing a bug which root cause is actually the route generation, but maybe it's better to have a safeguard here so that the controller is more resilient?

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -311,6 +326,8 @@ public function editAction($categoryId, Request $request)
                 'categoryId' => $categoryId,
+                'help_link' => $this->generateSidebarLink($request->attributes->get('_legacy_controller')),
+                'enableSidebar' => true,
                 'allowMenuThumbnailsUpload' => $editableCategory->canContainMoreMenuThumbnails(),
```
If the form type now handles the option `disable_menu_thumbnails_upload` is it still necessary to inject this in the template? Seems like it could be fetched from the form config directly no?

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,51 +71,87 @@ abstract class AbstractCategoryType extends TranslatorAwareType
      */
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $disableMenuThumbnailsUpload = false;
```
This part is quite foggy, I tried to understand it via the `canContainMoreMenuThumbnails` method and `MenuThumbnailId` but it lacks PHPdoc IMHO to understand its purpose You should add comments here and in the related methods to understand more the purpose of this

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,51 +71,87 @@ abstract class AbstractCategoryType extends TranslatorAwareType
+        if (isset($options['id_category'])) {
+            $disableMenuThumbnailsUpload = $options['disable_menu_thumbnails_upload'];
+        }
+        $genericCharactersHint = $this->trans('Invalid characters:', 'Admin.Global') . ' <>;=#{}';
```
The `<>;=#{}` should be stored in a const somewhere so it's reusable and accessible, and then you can use sprintf to generate this message

jolelievre replied:
```suggestion
        $genericCharactersHint = sprintf('%s %s', $this->trans('Invalid characters:', 'Admin.Global'), . TypedRegexValidator::CATALOG_CHARS);
```

jolelievre replied:
I'm not sure if `TypedRegexValidator::CATALOG_CHARS` is really adapted in this use case, but maybe one of its other constraints is. Regardless we should probably use this constraint here instead of a custom Regexp

jolelievre replied:
From the validator code it seems to match the regexp anyway:
https://github.com/PrestaShop/PrestaShop/blob/37f7fd4a9c5478c172263f4afec106f8686c199e/src/Core/ConstraintValidator/TypedRegexValidator.php#L108-L109

jolelievre replied:
So if you use this constraint you won't need to build the message manually

jolelievre commented on src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php:
```diff
@@ -67,51 +71,87 @@ abstract class AbstractCategoryType extends TranslatorAwareType
-        ConfigurationInterface $configuration
+        ConfigurationInterface $configuration,
+        Router $router,
+        CommandBusInterface $queryBus
```
Query bus should be injected recklessly into a form type And in this case especially, it doesn't seem to be used.

jolelievre commented on src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:
```diff
@@ -666,6 +666,8 @@ services:
       - '@prestashop.adapter.feature.multistore'
       - '@prestashop.adapter.legacy.configuration'
+      - '@router'
+      - '@prestashop.core.query_bus'
```
I believe this is not used, so it should be injected

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Categories/Blocks/form.html.twig:
```diff
@@ -34,96 +34,53 @@
-        </div>
-      </div>
-
+      {{ form_row(categoryForm.name) }}
```
The purpose of this simplified form is to simply have to use
`{{ form_row(categoryForm.name) }}`

Why do you need to display each field manually here? Is it because of the thumbnail special design? If so they should be customized either in the category form theme or maybe even in the UI kit if such display needs to be reused somewhere else

jolelievre replied:
Don't get me wrong it's already a lot more simplified, but I wonder what stops us to go to the whole thing

jolelievre reviewed (commented):
Thanks @JevgenijVisockij

Progi1984 commented:
Hi @JevgenijVisockij,
Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.
Thanks!
